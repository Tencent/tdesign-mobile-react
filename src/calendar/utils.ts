import { TCalendarValue } from './type';

export function getMonthByOffset(date: TCalendarValue, offset: number) {
  const innerDate = new Date(date);
  innerDate.setMonth(innerDate.getMonth() + offset);
  innerDate.setDate(1);
  return innerDate;
}

export function getYearByOffset(date: TCalendarValue, offset: number) {
  const innerDate = new Date(date);
  innerDate.setFullYear(innerDate.getFullYear() + offset);
  innerDate.setDate(1);
  return innerDate;
}

export const getPrevMonth = (date: TCalendarValue) => getMonthByOffset(date, -1);
export const getNextMonth = (date: TCalendarValue) => getMonthByOffset(date, 1);
export const getPrevYear = (date: TCalendarValue) => getYearByOffset(date, -1);
export const getNextYear = (date: TCalendarValue) => getYearByOffset(date, 1);
