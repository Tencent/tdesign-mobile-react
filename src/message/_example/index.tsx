import React from 'react';
import TDemoBlock from '../../../site/mobile/components/DemoBlock';
import TDemoHeader from '../../../site/mobile/components/DemoHeader';
import Base from './base';
import Theme from './theme';
import CloseAll from './close-all';
import './style/index.less';

export default function Demo() {
  return (
    <div className="tdesign-mobile-demo">
      <TDemoHeader title="Message 消息通知" summary="用于轻量级反馈或提示，不会打断用户操作。 " />
      <TDemoBlock title="01 类型" summary="弹窗内容为纯文本、标题和副标题、带输入框">
        <Base />
      </TDemoBlock>
      <TDemoBlock title="02 状态" summary="弹窗状态为普通弹窗、警示提示弹窗、成功提示弹窗、错误提示弹窗。">
        <Theme />
      </TDemoBlock>
      <TDemoBlock title="03 关闭所有通知" summary="关闭所有通知">
        <CloseAll />
      </TDemoBlock>
    </div>
  );
}
