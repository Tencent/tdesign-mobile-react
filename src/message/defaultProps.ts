/**
 * 该文件为脚本自动生成文件，请勿随意修改。如需修改请联系 PMC
 * */

import { TdMessageProps } from './type';

export const messageDefaultProps: TdMessageProps = {
  align: 'left',
  duration: 3000,
  icon: true,
  marquee: false,
  theme: 'info',
  defaultVisible: false,
};
