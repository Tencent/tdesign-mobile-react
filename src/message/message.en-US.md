:: BASE_DOC ::

## API

### Message Props

name | type | default | description | required
-- | -- | -- | -- | --
className | String | - | className of component | N
style | Object | - | CSS(Cascading Style Sheets)，Typescript：`React.CSSProperties` | N
align | String | left | options: left/center。Typescript：`MessageAlignType` `type MessageAlignType = 'left' \| 'center'`。[see more ts definition](https://github.com/Tencent/tdesign-mobile-react/tree/develop/src/message/type.ts) | N
closeBtn | TNode | - | Typescript：`string \| boolean \| TNode`。[see more ts definition](https://github.com/Tencent/tdesign-mobile-react/blob/develop/src/common.ts) | N
content | TNode | - | Typescript：`string \| TNode`。[see more ts definition](https://github.com/Tencent/tdesign-mobile-react/blob/develop/src/common.ts) | N
duration | Number | 3000 | \- | N
icon | TNode | true | Typescript：`boolean \| TNode`。[see more ts definition](https://github.com/Tencent/tdesign-mobile-react/blob/develop/src/common.ts) | N
link | TNode | - | Typescript：`string \| object \| TNode`。[see more ts definition](https://github.com/Tencent/tdesign-mobile-react/blob/develop/src/common.ts) | N
marquee | Boolean / Object | false | Typescript：`boolean \| MessageMarquee` `interface MessageMarquee { speed?: number; loop?: number; delay?: number }`。[see more ts definition](https://github.com/Tencent/tdesign-mobile-react/tree/develop/src/message/type.ts) | N
offset | Array | - | Typescript：`Array<string \| number>` | N
theme | String | info | options: info/success/warning/error。Typescript：`MessageThemeList` `type MessageThemeList = 'info' \| 'success' \| 'warning' \| 'error'`。[see more ts definition](https://github.com/Tencent/tdesign-mobile-react/tree/develop/src/message/type.ts) | N
visible | Boolean | false | \- | N
defaultVisible | Boolean | false | uncontrolled property | N
zIndex | Number | - | \- | N
onChange | Function |  | Typescript：`(visible: boolean) => void`<br/>`deprecated` | N
onClose | Function |  | Typescript：`(context: { trigger: 'close-click' \| 'duration-end', e?: MouseEvent }) => void`<br/>`deprecated`。close message event。Typescript：`string \| TNode`。[see more ts definition](https://github.com/Tencent/tdesign-mobile-react/blob/develop/src/common.ts) | N
onCloseBtnClick | Function |  | Typescript：`(context: { e: MouseEvent }) => void`<br/> | N
onClosed | Function |  | Typescript：`() => void`<br/>`deprecated` | N
onDurationEnd | Function |  | Typescript：`() => void`<br/> | N
onLinkClick | Function |  | Typescript：`(context: { e: MouseEvent }) => void`<br/> | N
onOpen | Function |  | Typescript：`() => void`<br/>`deprecated` | N
onOpened | Function |  | Typescript：`() => void`<br/>`deprecated` | N
