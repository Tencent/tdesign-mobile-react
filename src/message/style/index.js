// eslint-disable-next-line import/no-relative-packages
import '../../_common/style/mobile/components/message/v2/_index.less';
