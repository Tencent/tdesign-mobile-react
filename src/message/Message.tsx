import React, { useEffect, useState, useRef } from 'react';
import classNames from 'classnames';
import { useTimeout } from 'ahooks';
import { isObject } from 'lodash-es';
import { CSSTransition } from 'react-transition-group';
import { Icon } from 'tdesign-icons-react';

import type { TdMessageProps, MessageMarquee } from './type';
import useMessageCssTransition from './hooks/useMessageCssTransition';
import useDefaultProps from '../hooks/useDefaultProps';
import { usePrefixClass } from '../hooks/useClass';

import type { StyledProps } from '../common';
import parseTNode from '../_util/parseTNode';
import { convertUnit } from '../_util/convertUnit';
import { messageDefaultProps } from './defaultProps';

import Link from '../link';

export interface MessageProps extends TdMessageProps, StyledProps {
  container?: Element;
}

interface IInnerState {
  duration: number;
  offset: number;
  listWidth: number;
  itemWidth: number;
  scroll: {
    marquee: boolean;
    speed: number;
    loop: number;
    delay: number;
  };
}

const changeNumToStr = (arr: TdMessageProps['offset'] = []) => arr.map((item) => convertUnit(item));

const getMessageStylesOffset = (offset: TdMessageProps['offset']) => {
  const arr = changeNumToStr(offset);
  return offset
    ? {
        top: arr[0],
        right: arr[1],
        left: arr[1],
      }
    : {};
};

const Message: React.FC<MessageProps> = (originProps) => {
  const props = useDefaultProps(originProps, messageDefaultProps);
  const {
    className,
    style,
    children,
    align,
    closeBtn,
    duration,
    theme,
    marquee,
    visible,
    defaultVisible,
    link,
    zIndex,
    content,
    icon,
    container,
    onDurationEnd,
    onCloseBtnClick,
  } = props;

  const name = usePrefixClass('message');

  /**
   * 获取visibleChange函数引用
   */
  const scrollingHandler = useRef<Function>(null);

  /**
   * duration为0时，取消倒计时
   */
  const [messageDuration] = useState<number | null>(duration || null);
  const [state, setState] = useState<IInnerState>({
    duration: 0,
    offset: 0,
    listWidth: 0,
    itemWidth: 0,
    scroll: {
      marquee: false,
      speed: 50,
      loop: -1, // 值为 -1 表示循环播放，值为 0 表示不循环播放
      delay: 300,
    },
  });

  const [messageVisible, setVisible] = useState<boolean>(visible ?? defaultVisible);

  useEffect(() => {
    if (typeof visible !== 'undefined' && messageVisible !== visible) {
      setVisible(visible);
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [visible]);

  const contentRef = useRef<HTMLDivElement>(null);
  const textWrapRef = useRef<HTMLDivElement>(null);
  const textRef = useRef<HTMLDivElement>(null);

  const cssTransitionState = useMessageCssTransition({
    contentRef,
    classPrefix: 'message',
    container: container as HTMLElement,
  });

  useTimeout(() => {
    setVisible(false);
    onDurationEnd?.();
  }, messageDuration);

  const handleScrolling = () => {
    if (!marquee || (isObject(marquee) && (marquee as MessageMarquee))?.loop === 0) {
      return;
    }

    const { loop, speed, delay } = state.scroll;

    const scroll = {
      marquee: true,
      // 负数统一当作循环播放
      loop: Math.max(
        props.marquee === true || (props.marquee as MessageMarquee)?.loop == null
          ? loop
          : (props.marquee as MessageMarquee)?.loop,
        -1,
      ),
      // 速度必须为正数
      speed: Math.max(
        props.marquee === true || (props.marquee as MessageMarquee)?.speed == null
          ? speed
          : (props.marquee as MessageMarquee)?.speed,
        1,
      ),
      // 延迟不可为负数
      delay: Math.max(
        props.marquee === true || (props.marquee as MessageMarquee)?.delay == null
          ? delay
          : (props.marquee as MessageMarquee)?.delay,
        0,
      ),
    };
    setState({
      ...state,
      scroll,
      offset: 0,
    });

    // 设置动画
    setTimeout(() => {
      const textWrapDOMWidth = textWrapRef.current?.getBoundingClientRect().width;
      const textDOMWidth = textRef.current?.getBoundingClientRect().width;

      setState({
        ...state,
        scroll,
        offset: -textDOMWidth,
        duration: textDOMWidth / state.scroll.speed,
        listWidth: textWrapDOMWidth,
        itemWidth: textDOMWidth,
      });
    }, state.scroll.delay);
  };

  scrollingHandler.current = handleScrolling;

  useEffect(() => {
    scrollingHandler.current();
  }, []);

  const animateStyle = {
    transform: state.offset ? `translateX(${state.offset}px)` : '',
    transitionDuration: `${state.duration}s`,
    transitionTimingFunction: 'linear',
  };

  const resetTransition = () => {
    setState({
      ...state,
      duration: 0,
      offset: state.listWidth,
    });

    setTimeout(() => {
      setState({
        ...state,
        offset: -state.itemWidth,
        duration: (state.itemWidth + state.listWidth) / state.scroll.speed,
      });
    }, 0);
  };

  // 动画结束后，初始化动画
  const handleTransitionend = () => {
    resetTransition();

    if (state.scroll.loop === -1) {
      return;
    }

    const newState = {
      ...state,
      scroll: {
        ...state.scroll,
        loop: -(-state.scroll.loop),
      },
    };

    if (state.scroll.loop === 0) {
      newState.scroll.marquee = false;
    }

    setState(newState);
  };

  const onLinkClick = (e) => {
    props.onLinkClick?.(e);
  };
  const leftIcon = parseTNode(
    icon,
    null,
    <Icon name={`${theme === 'success' ? 'check' : 'error'}-circle-filled`} size={22} />,
  );

  const getLinkContent = () => {
    if (typeof link === 'string') {
      return <Link content={link} />;
    }
    if (isObject(link)) {
      return <Link theme="primary" {...link} />;
    }
    return parseTNode(link);
  };

  const clickCloseButton = (e) => {
    setVisible(false);
    onCloseBtnClick?.(e);
  };

  const closeButton =
    closeBtn === true ? (
      <Icon className={`${name}--close-btn`} name="close" size={22} onClick={clickCloseButton} />
    ) : (
      closeBtn
    );

  return (
    <CSSTransition in={messageVisible} appear {...cssTransitionState.props} unmountOnExit>
      <div
        className={classNames(`${name}`, className, `${name}--${theme}`, { [`${name}-align--${align}`]: !!align })}
        ref={contentRef}
        style={{ zIndex, ...getMessageStylesOffset(props.offset), ...style }}
      >
        <div className={`${name}__icon--left`}>{leftIcon}</div>
        <div ref={textWrapRef} className={classNames(`${name}__text-wrap`, { [`${name}__text-nowrap`]: marquee })}>
          <div
            ref={textRef}
            className={`${name}__text`}
            style={state.scroll.marquee ? animateStyle : {}}
            onTransitionEnd={handleTransitionend}
          >
            {parseTNode(content) || parseTNode(children)}
          </div>
        </div>
        {link && (
          <div className={`${name}__link`} onClick={onLinkClick}>
            {getLinkContent()}
          </div>
        )}
        {closeButton}
      </div>
    </CSSTransition>
  );
};

Message.displayName = 'Message';

export default Message;
