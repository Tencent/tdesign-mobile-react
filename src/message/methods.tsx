import React from 'react';
import ReactDOM from 'react-dom';
import { MessageThemeList, MessageActionOptionsType } from './type';
import { messageDefaultProps } from './defaultProps';
import Message from './Message';

const isBrowser = typeof window !== 'undefined';

const instanceMap: Map<Element, Record<string, Element>> = new Map();

function destroy(context: Element, root: Element) {
  if (context.contains(root)) {
    context.removeChild(root);
    if (instanceMap.has(root)) {
      instanceMap.delete(root);
    }
  }
}

const createMessage = (props, theme?: MessageThemeList) => {
  const config = {
    ...messageDefaultProps,
    visible: true,
    context: isBrowser ? document.body : null,
    ...props,
  };
  const { context } = config;
  if (!context) {
    console.error('未找到组件, 请确认 context 是否正确');
    return;
  }

  const root = document.createElement('div');
  context.appendChild(root);

  instanceMap.set(root, {
    context,
  });

  ReactDOM.render(<Message {...{ ...config, theme, container: root }} />, root);
};

const closeAll = () => {
  if (instanceMap instanceof Map) {
    for (const [key, value] of instanceMap) {
      const { context } = value;
      destroy(context as Element, key);
    }
  }
};

export default {
  info: (props: MessageActionOptionsType) => createMessage(props, 'info'),
  success: (props: MessageActionOptionsType) => createMessage(props, 'success'),
  warning: (props: MessageActionOptionsType) => createMessage(props, 'warning'),
  error: (props: MessageActionOptionsType) => createMessage(props, 'error'),
  closeAll,
};
