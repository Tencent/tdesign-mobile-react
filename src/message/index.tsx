import React from 'react';
import { attachMethodsToComponent } from '../_util/attachMethodsToComponent';
import _Message from './Message';
import type { MessageProps } from './Message';
import * as methods from './methods';
import type {
  MessageInfoMethod,
  MessageErrorMethod,
  MessageCloseAll,
  MessageSuccessMethod,
  MessageWarningMethod,
} from './type';

import './style';

export * from './type';
export interface MessageType extends React.FC<MessageProps> {
  info: MessageInfoMethod;
  warning: MessageWarningMethod;
  success: MessageSuccessMethod;
  error: MessageErrorMethod;
  closeAll: MessageCloseAll;
}

const { info, warning, success, error, closeAll } = methods.default;

export const Message: MessageType = attachMethodsToComponent(_Message, {
  info,
  warning,
  success,
  error,
  closeAll,
});

export default Message;
