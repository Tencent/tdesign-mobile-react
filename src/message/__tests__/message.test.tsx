import { act, beforeEach, describe, expect, fireEvent, render, test, vi } from '@test/utils';
import React from 'react';
import { MessageProps } from 'tdesign-mobile-react/message/Message';
import Message, { MessageAlignType } from '../index';

const classNamePrefix = 't';
const messageClassName = `${classNamePrefix}-message`;
const querySelectorName = `.${messageClassName}`;

function getClassNameWithPrefix(className: string) {
  return `${classNamePrefix}-${className}`;
}

describe('Message', () => {
  beforeEach(() => {
    vi.useFakeTimers();
  });

  describe('props', () => {
    test('className', () => {
      const customClassName = 'custom-className';
      const { container } = render(<Message className={customClassName} visible={true} />);
      expect(container.querySelector(querySelectorName)).toHaveClass(customClassName);
    });

    test('style', () => {
      const customStyle = { color: 'rgb(255, 0, 0)' };
      const { container } = render(<Message style={customStyle} visible={true} />);
      expect(container.querySelector(querySelectorName)).toHaveStyle(customStyle);
    });

    test('align', () => {
      const defaultAlign = 'left';
      {
        // defaultAlign
        const { container } = render(<Message visible={true} />);
        expect(container.querySelector(querySelectorName)).toHaveClass(
          getClassNameWithPrefix(`message-align--${defaultAlign}`),
        );
      }

      {
        // ‘center'
        const centerAlign: MessageAlignType = 'center';
        const { container } = render(<Message visible={true} align={centerAlign} />);
        expect(container.querySelector(querySelectorName)).toHaveClass(
          getClassNameWithPrefix(`message-align--${centerAlign}`),
        );
      }

      {
        // 'left'
        const leftAlign: MessageAlignType = 'left';
        const { container: customAlignContainer } = render(<Message visible={true} align={leftAlign} />);
        expect(customAlignContainer.querySelector(querySelectorName)).toHaveClass(
          getClassNameWithPrefix(`message-align--${leftAlign}`),
        );
      }
    });

    test('closeBtn', () => {
      const closeBtnClassNameSelector = `.${getClassNameWithPrefix(`message--close-btn`)}`;
      // Default closeBtn is undefined
      {
        const { container } = render(<Message visible={true} />);
        expect(container.querySelector(closeBtnClassNameSelector)).toBeNull();
      }

      // Show closeBtn
      {
        const { container } = render(<Message visible={true} closeBtn={true} />);
        expect(container.querySelector(closeBtnClassNameSelector)).toBeTruthy();
      }

      // string
      {
        const closeBtnText = 'closeBtnText';
        const { container } = render(<Message visible={true} closeBtn={closeBtnText} />);
        expect(container.querySelector(querySelectorName).textContent).toBe(closeBtnText);
      }

      // TNode
      {
        const closeBtnClassname = 'close-btn';
        const closeBtnClickHandler = vi.fn();
        const { container } = render(
          <Message
            visible={true}
            closeBtn={
              <div className={closeBtnClassname} onClick={closeBtnClickHandler}>
                closeBtnText
              </div>
            }
          />,
        );
        const closeBtn = container.querySelector(`.${closeBtnClassname}`);

        expect(closeBtn).toBeTruthy();

        fireEvent.click(closeBtn);
        expect(closeBtnClickHandler).toBeCalledTimes(1);
      }
    });

    test('content', () => {
      // string
      {
        const content = 'content';
        const { container } = render(<Message content={content} visible={true} />);
        expect(container.querySelector(`.${getClassNameWithPrefix('message__text')}`).textContent).toBe(content);
      }

      // TNode
      {
        const customContentName = 'custom-content';
        const customContent = 'customContent';
        const clickHandler = vi.fn();

        const { container } = render(
          <Message
            content={
              <div className={customContentName} onClick={clickHandler}>
                {customContent}
              </div>
            }
            visible={true}
          />,
        );
        const customContentDiv = container.querySelector(`.${customContentName}`);
        expect(customContentDiv.textContent).toBe(customContent);

        fireEvent.click(customContentDiv);
        expect(clickHandler).toBeCalledTimes(1);
      }
    });

    test('duration', async () => {
      const defaultDuration = 3000;

      {
        // Default duration
        const { container } = render(<Message visible={true} />);
        await act(async () => {
          vi.advanceTimersByTime(defaultDuration - 1);
        });
        expect(container.querySelector(querySelectorName)).toHaveClass('message-enter-active');
        await act(async () => {
          vi.advanceTimersByTime(1);
        });
        expect(container.querySelector(querySelectorName)).toHaveClass('message-leave');
      }

      {
        // Custom duration
        const duration = 1000;
        const { container } = render(<Message visible={true} duration={duration} />);
        await act(async () => {
          vi.advanceTimersByTime(duration);
        });
        expect(container.querySelector(querySelectorName)).toHaveClass('message-leave');
      }

      {
        // Disabled duration
        const duration = 0;
        const { container } = render(<Message visible={true} duration={duration} />);
        await act(async () => {
          vi.advanceTimersByTime(defaultDuration * 2);
        });
        expect(container.querySelector(querySelectorName)).toHaveClass('message-enter-active');
      }
    });

    test('link', () => {
      {
        // string
        const linkContent = '链接';
        const { container } = render(<Message visible={true} link={linkContent} />);
        expect(container.querySelector(`.${getClassNameWithPrefix('link__content')}`).textContent).toBe(linkContent);
      }

      {
        // Use link params
        const linkParams = {
          size: 'small',
          theme: 'danger',
          content: 'link content',
        };
        // todo PM fix this link 描述有问题
        const { container } = render(<Message visible={true} link={linkParams} />);
        expect(container.querySelector(`.${getClassNameWithPrefix('link--danger')}`)).toBeTruthy();
      }

      {
        // TNode - function
        const customContentName = 'custom-content';
        const clickHandler = vi.fn();
        const { container } = render(
          <Message
            visible={true}
            link={() => (
              <div onClick={clickHandler} className={customContentName}>
                customContent
              </div>
            )}
          />,
        );

        const customContentDiv = container.getElementsByClassName(customContentName)[0];
        expect(customContentDiv).toBeTruthy();

        fireEvent.click(customContentDiv);
        expect(clickHandler).toBeCalledTimes(1);
      }

      {
        // TNode - Element
        const customContentName = 'custom-content';
        const clickHandler = vi.fn();
        const { container } = render(
          <Message
            visible={true}
            link={
              <div onClick={clickHandler} className={customContentName}>
                customContent
              </div>
            }
          />,
        );

        const customContentDiv = container.getElementsByClassName(customContentName)[0];
        expect(customContentDiv).toBeTruthy();

        fireEvent.click(customContentDiv);
        expect(clickHandler).toBeCalledTimes(1);
      }
    });

    test('marquee', async () => {
      const textWrapRect = { width: 200 } as DOMRect;
      const textRect = { width: 300 } as DOMRect;

      const originalGetBoundingClientRect = Element.prototype.getBoundingClientRect;
      Element.prototype.getBoundingClientRect = function () {
        const className = (this as Element).className || '';
        if (typeof className === 'string') {
          if (className.includes(getClassNameWithPrefix('message__text-wrap'))) return textWrapRect as any;
          if (className.includes(getClassNameWithPrefix('message__text'))) return textRect as any;
        }
        return originalGetBoundingClientRect.call(this);
      } as any;

      {
        // Boolean
        const { container } = render(<Message visible={true} marquee={true} content="hello marquee" />);

        const textEl = container.querySelector(`.${getClassNameWithPrefix('message__text')}`) as HTMLDivElement;
        expect(textEl).toBeTruthy();

        // After delay, first play: snapshot styles applied
        await act(async () => {
          vi.advanceTimersByTime(100);
        });
        expect(textEl).toMatchSnapshot();

        // Fire transitionend -> snapshot reset state
        textEl.dispatchEvent(new Event('transitionend'));
        expect(textEl).toMatchSnapshot();

        // Next play after delay (playedCount > 0): snapshot styles applied
        await act(async () => {
          vi.advanceTimersByTime(10);
        });
        expect(textEl).toMatchSnapshot();
      }

      {
        // Object
        const { container } = render(
          <Message visible={true} marquee={{ loop: 2, speed: 100, delay: 100 }} content="hello marquee" />,
        );

        const textEl = container.querySelector(`.${getClassNameWithPrefix('message__text')}`) as HTMLDivElement;
        expect(textEl).toBeTruthy();

        // After delay, first play: snapshot styles applied
        await act(async () => {
          vi.advanceTimersByTime(100);
        });
        expect(textEl).toMatchSnapshot();

        // Fire transitionend -> snapshot reset state
        textEl.dispatchEvent(new Event('transitionend'));
        expect(textEl).toMatchSnapshot();

        // Next play after delay (playedCount > 0): snapshot styles applied
        await act(async () => {
          vi.advanceTimersByTime(10);
        });
        expect(textEl).toMatchSnapshot();
      }

      // Cleanup mock
      Element.prototype.getBoundingClientRect = originalGetBoundingClientRect;
    });

    test('offset', () => {
      const { container } = render(<Message visible={true} offset={[-100, 100]} />);
      expect(container.querySelector(querySelectorName)).toHaveStyle({ top: '-100px', left: '100px' });
    });

    test('theme', () => {
      {
        // Default theme
        const { container } = render(<Message visible={true} />);
        const iconContainer = container.querySelector(`.${getClassNameWithPrefix('message__icon--left')}`);
        expect(iconContainer.querySelector('svg')).toBeTruthy();
      }

      {
        // All themes
        const themes: MessageProps['theme'][] = ['info', 'success', 'warning', 'error'];
        themes.forEach((theme) => {
          const { container } = render(<Message theme={theme} visible={true} />);
          const iconContainer = container.querySelector(`.${getClassNameWithPrefix('message__icon--left')}`);
          expect(iconContainer.querySelector('svg')).toBeTruthy();
        });
      }
    });

    test('visible', async () => {
      // Default visible: false
      {
        const { container } = render(<Message />);
        expect(container.querySelector(querySelectorName)).toBeNull();
      }

      // Visible false
      {
        const { container } = render(<Message visible={false} />);
        expect(container.querySelector(querySelectorName)).toBeNull();
      }

      // Control visible
      {
        const { container, rerender } = render(<Message visible={true} />);
        const messageElement = container.querySelector(querySelectorName);
        expect(messageElement).toHaveClass('message-enter-active');

        await act(async () => {
          rerender(<Message visible={false} />);
        });
        expect(messageElement).toHaveClass('message-leave');
      }
    });

    test('defaultVisible', async () => {
      {
        // True
        const { container } = render(<Message defaultVisible={true} />);
        expect(container.querySelector(querySelectorName)).toBeTruthy();
      }

      {
        // False
        const { container } = render(<Message defaultVisible={false} />);
        expect(container.querySelector(querySelectorName)).toBeNull();
      }

      {
        // Combine with visible
        const { container } = render(<Message defaultVisible={true} visible={false} />);
        expect(container.querySelector(querySelectorName)).toBeNull();
      }

      {
        // Combine with visible
        const { container } = render(<Message defaultVisible={true} visible={true} />);
        expect(container.querySelector(querySelectorName)).toBeTruthy();
      }

      {
        // Combine with visible
        const { container } = render(<Message defaultVisible={false} visible={false} />);
        expect(container.querySelector(querySelectorName)).toBeNull();
      }

      {
        // Combine with visible
        const { container } = render(<Message defaultVisible={false} visible={true} />);
        expect(container.querySelector(querySelectorName)).toBeTruthy();
      }

      {
        // Change should no effect
        const { container, rerender } = render(<Message defaultVisible={true} />);
        rerender(<Message defaultVisible={false} />);
        expect(container.querySelector(querySelectorName)).toBeTruthy();
      }
    });

    test('zIndex', () => {
      // todo PM fix this: zIndex default should be 5000
      // const defaultZIndex = '5000';
      // {
      //   const { container } = render(<Message visible={true} />);
      //   expect(window.getComputedStyle(container.querySelector(querySelectorName)).zIndex).toBe(defaultZIndex);
      // }

      // Custom zIndex
      const { container } = render(<Message zIndex={1000} visible={true} />);
      expect(container.querySelector(querySelectorName)).toHaveStyle({ zIndex: 1000 });
    });

    test('onCloseBtnClick', async () => {
      {
        // When closeBtn is TNode
        const closeBtnClickHandler = vi.fn();
        const elementClickHandler = vi.fn();

        const closeBtnClassName = 'close-btn';
        const closeBthContent = (
          <div className={closeBtnClassName} onClick={elementClickHandler}>
            close
          </div>
        );

        const { container } = render(
          <Message visible={true} onCloseBtnClick={closeBtnClickHandler} closeBtn={closeBthContent} />,
        );
        const button = container.querySelector(`.${closeBtnClassName}`);

        fireEvent.click(button);
        expect(closeBtnClickHandler).toBeCalledTimes(1);
        expect(elementClickHandler).toBeCalledTimes(1);
      }

      {
        // Validate event structure
        const onCloseBtnClick = vi.fn<(context: { e: MouseEvent }) => void>();

        const closeBtnClassName = 'close-btn';
        const closeBthContent = <div className={closeBtnClassName}>close</div>;

        const { container } = render(
          <Message visible={true} closeBtn={closeBthContent} onCloseBtnClick={onCloseBtnClick as any}>
            测试内容
          </Message>,
        );

        const closeButton = container.querySelector(`.${closeBtnClassName}`);

        fireEvent.click(closeButton);

        const syntheticEvent = onCloseBtnClick.mock.calls[0][0].e;
        expect((syntheticEvent as any).nativeEvent).toBeInstanceOf(MouseEvent);
      }

      {
        // When closeBtn is Boolean
        const closeBtnClickHandler = vi.fn();
        const { container } = render(<Message visible={true} onCloseBtnClick={closeBtnClickHandler} closeBtn={true} />);
        const button = container.querySelector(`.${messageClassName}__close-wrap`);

        fireEvent.click(button);
        expect(closeBtnClickHandler).toBeCalledTimes(1);
      }
    });

    test('onDurationEnd', async () => {
      const durationTime = 3000;
      const onDurationEnd = vi.fn();

      render(<Message visible={true} duration={durationTime} onDurationEnd={onDurationEnd} />);

      await act(async () => {
        vi.advanceTimersByTime(durationTime - 1);
      });

      expect(onDurationEnd).toBeCalledTimes(0);

      await act(async () => {
        vi.advanceTimersByTime(1);
      });

      expect(onDurationEnd).toBeCalledTimes(1);
    });

    test('onLinkClick', async () => {
      {
        // When link is string
        const linkClickHandler = vi.fn();
        const { container } = render(<Message visible={true} onLinkClick={linkClickHandler} link="link" />);
        const button = container.querySelector(`.${getClassNameWithPrefix('link')}`);

        fireEvent.click(button);
        expect(linkClickHandler).toBeCalledTimes(1);
      }

      {
        // Where link is object
        const linkClickHandler = vi.fn();
        const linkParams = {
          size: 'small',
          theme: 'danger',
          content: 'link content',
        };

        const { container } = render(
          <Message visible={true} onLinkClick={linkClickHandler} link={linkParams as any} />,
        );
        const button = container.querySelector(`.${getClassNameWithPrefix('link')}`);

        fireEvent.click(button);
        expect(linkClickHandler).toBeCalledTimes(1);
      }

      {
        // When link is TNode
        const linkClickHandler = vi.fn();
        const elementClickHandler = vi.fn();

        const linkClassName = 'link-classname';
        const linkContent = (
          <div className={linkClassName} onClick={elementClickHandler}>
            link
          </div>
        );

        const { container } = render(<Message visible={true} onLinkClick={linkClickHandler} link={linkContent} />);
        const button = container.querySelector(`.${linkClassName}`);

        fireEvent.click(button);
        expect(linkClickHandler).toBeCalledTimes(1);
        expect(elementClickHandler).toBeCalledTimes(1);
      }

      {
        // Validate event structure
        const onLinkClick = vi.fn<(context: { e: MouseEvent }) => void>();

        const linkClassName = 'link-classname';
        const linkContent = <div className={linkClassName}>link</div>;

        const { container } = render(<Message visible={true} onLinkClick={onLinkClick as any} link={linkContent} />);

        const closeButton = container.getElementsByClassName(linkClassName)[0];

        fireEvent.click(closeButton);

        const syntheticEvent = onLinkClick.mock.calls[0][0].e;
        expect((syntheticEvent as any).nativeEvent).toBeInstanceOf(MouseEvent);
      }
    });

    test('children', () => {
      // string
      {
        const content = 'content';
        const { container } = render(<Message visible={true}>{content}</Message>);
        expect(container.querySelector(`.${getClassNameWithPrefix('message__text')}`).textContent).toBe(content);
      }

      // TNode
      {
        const customContentName = 'custom-content';
        const customContent = 'customContent';
        const clickHandler = vi.fn();

        const { container } = render(
          <Message
            content={
              <div className={customContentName} onClick={clickHandler}>
                {customContent}
              </div>
            }
            visible={true}
          />,
        );
        const customContentDiv = container.querySelector(`.${customContentName}`);
        expect(customContentDiv.textContent).toBe(customContent);

        fireEvent.click(customContentDiv);
        expect(clickHandler).toBeCalledTimes(1);
      }
    });
  });
});
