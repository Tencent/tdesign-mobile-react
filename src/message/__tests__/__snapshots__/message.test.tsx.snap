// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Message > props > marquee 1`] = `
<div
  class="t-message__text"
>
  hello marquee
</div>
`;

exports[`Message > props > marquee 2`] = `
<div
  class="t-message__text"
  style="transform: translateX(500px); transition-duration: 0s;"
>
  hello marquee
</div>
`;

exports[`Message > props > marquee 3`] = `
<div
  class="t-message__text"
  style="transform: translateX(500px); transition-duration: 0s;"
>
  hello marquee
</div>
`;

exports[`Message > props > marquee 4`] = `
<div
  class="t-message__text"
  style="transform: translateX(-300px); transition-duration: 3s; transition-timing-function: linear;"
>
  hello marquee
</div>
`;

exports[`Message > props > marquee 5`] = `
<div
  class="t-message__text"
  style="transform: translateX(500px); transition-duration: 0s; transition-timing-function: linear;"
>
  hello marquee
</div>
`;

exports[`Message > props > marquee 6`] = `
<div
  class="t-message__text"
  style="transform: translateX(500px); transition-duration: 0s; transition-timing-function: linear;"
>
  hello marquee
</div>
`;
